### ISSUE TITLE
Setting iframe.srcdoc doesn't work in WeChat browser.

### ISSUE THREAD

**Issue by shawn0326** [guybedford/es-module-shims#329]
The 1.5.16 version does not work properly in WeChat's browser (the webview in WeChat app 8.0.26). I tried to find the reason and found that after iframe.srcdoc was set, the iframe did not load normally and returned the message event.

I changed some code and it works fine now, but maybe there is a better way.

````javascript
    // setting srcdoc is not supported in wechat browser
    const weChat = !!navigator.userAgent.match(/MicroMessenger/i);

    if (!hasDocument || weChat)
      return Promise.all([
        supportsImportMaps || dynamicImport(createBlob(importMetaCheck)).then(() => supportsImportMeta = true, noop),
        cssModulesEnabled && dynamicImport(createBlob(cssModulesCheck.replace('x', createBlob('', 'text/css')))).then(() => supportsCssAssertions = true, noop),
        jsonModulesEnabled && dynamicImport(createBlob(jsonModulescheck.replace('x', createBlob('{}', 'text/json')))).then(() => supportsJsonAssertions = true, noop),
      ]);
````

**Comment by guybedford** [comment]
This [check](https://github.com/guybedford/es-module-shims/blob/main/src/features.js#L47):

```js
if ('srcdoc' in iframe)
```

is supposed to detect is srcdoc is supported, apparently that's not working in WeChat's browser? I don't have a testing environment to test this, could you perhaps let me know if a variation of this test, perhaps something like:

```js
const iframe = document.createElement('iframe');
const html = '<h1>html</h1>';
iframe.srcdoc = html;
const supportsSrcDoc = iframe.srcdoc.length === html.length;
console.log(supportsSrcDoc);
```

works as a feature detection? Feature detections are always preferable because user agents can vary a lot.

**Comment by shawn0326** [comment]
I tested the above code in wechat browser and it returns true normally:
````javascript
alert(supportsSrcDoc); // return true
````

However, I also did a test like this:
````javascript
  const iframe = document.createElement('iframe');
  const html = '<h1>html</h1>';
  iframe.srcdoc = html;
  const supportsSrcDoc = iframe.srcdoc.length === html.length;
  alert(supportsSrcDoc); // alert true
  document.head.appendChild(iframe);
  iframe.onload = function() { alert ('iframe loaded') }
````
but iframe's onload event doesn't fire.

I also tried adding a script to the iframe, but it didn't execute
````javascript
const html = '<script>alert(2)</script>';
````

In chrome and my mobile browser, the above code can be executed normally.

**Comment by shawn0326** [comment]
It looks like the iframe setup is all successful, but the page inside the iframe is not executed. 
This may be a limitation of the WeChat browser, but it does make the page referencing es-module-shims.js stuck.

**Comment by guybedford** [comment]
@shawn0326 thanks so much for your help in testing some of these cases. If you're able to try a few more things that would help a lot, otherwise I can try and work out how to get a debugging environment as well at some point when I have some time. 

Another test would be like what we used to do - if you add a script, can you see the script is there?

For example:

```js
 const iframe = document.createElement('iframe');
const html = `<script><${''}/script>`;
iframe.srcdoc = html;
// attach onload _before_ adding to head
iframe.onload = function () {
  console.log(iframe.contentDocument);
};
document.head.appendChild(iframe);
console.log(iframe.contentDocument);
```

Is the content document defined in both cases? And does the logged document include the `<script>` DOM element in either case?

**Comment by shawn0326** [comment]
> ```js
> const iframe = document.createElement('iframe');
> const html = `<script><${''}/script>`;
> iframe.srcdoc = html;
> // attach onload _before_ adding to head
> iframe.onload = function () {
>   console.log(iframe.contentDocument);
> };
> document.head.appendChild(iframe);
> console.log(iframe.contentDocument);
> ```

At present, I haven't found a way to open the browser's console, so I can only use alert instead of console.log to test.
The onload event is still not triggered, iframe.contentDocument can be alerted, but it is displayed as [object HTMLDocument]. Don't know if there is any way to see its internal structure through alert, if necessary.

**Comment by guybedford** [comment]
@shawn0326 can you try this version:

```js
const iframe = document.createElement('iframe');
const html = `<script>alert("script")<${''}/script>`;
iframe.addEventListener('load', () => {
  alert("loaded");
}, false);
document.head.appendChild(iframe);
iframe.srcdoc = html;
```

and let me know what alerts you get?

**Comment by shawn0326** [comment]
> and let me know what alerts you get?

I can get "loaded" now, but can't get "script".

**Comment by guybedford** [comment]
I guess the question then is if WeChat is blocking all srcdoc scripts, or just an alert from the iframe?

Perhaps try:

```js
const iframe = document.createElement('iframe');
const html = `<script>parent.postMessage('msg', '*');<${''}/script>`;
iframe.addEventListener('load', () => {
  alert("IFrame loaded");
}, false);
window.addEventListener('message', () => {
  alert("IFrame message");
}, false);
document.head.appendChild(iframe);
iframe.srcdoc = html;
```

**Comment by guybedford** [comment]
Also worth trying with an error listener as well:

```js
iframe.onerror = function (e) {
  alert("iframe error - " + e.toString());
}
```

**Comment by guybedford** [comment]
I've updated the example with error and HTML logging, which should cover everything now.

**Comment by shawn0326** [comment]
@guybedford 

wechat browser alerts: 
1. IFrame loaded
2. <html><head></head><body></body></html>
3. IFrame message - msg2

chrome alerts:
1. IFrame loaded
2. <html><head></head><body></body></html>
3. IFrame loaded
4. <html><head><script>parent.postMessage('msg', '*');</script></head><body></body></html>
5. IFrame message - msg
6. IFrame message - msg2

i also tested in ios safari, `msg2` alert twice:
1. IFrame loaded
2. <html><head></head><body></body></html>
3. IFrame message - msg2
4. IFrame loaded
5. <html><head><script>parent.postMessage('msg', '*');</script></head><body></body></html>
6. IFrame message - msg
7. IFrame message - msg2

**Comment by guybedford** [comment]
Thanks again @shawn0326 for the thorough testing here! Using these results I've been able to put together a reliable detection and fallback in https://github.com/guybedford/es-module-shims/pull/330. Please test out the PR if you can, otherwise I'll aim to land the fix in the next few days.

**Comment by shawn0326** [comment]
Aha！This worked in the wechat browser, great job！

**Comment by guybedford** [comment]
You're welcome, released in 1.5.17.
